src/extension.ts

```diff
@@ -89,44 +90,88 @@ const pluralDecoration = vscode.window.createTextEditorDecorationType({
 		return;
 	}
-	let colorMap = new Map<string, vscode.Range[]>();
+	let currentLevel = 0;
+	let isInPlaceholders = -1;
+	let isInMetadata = -1;
 	visit(editor.document.getText(), {
 		onLiteralValue: (value: string, offset: number) => {
-			let openBrackets: number[] = [];
-			for (let index = 0; index < value.length; index++) {
-				const char = value.charAt(index);
-				if (char === '{') {
-					openBrackets.push(index);
-				} else if (char === '}') {
-					let start = openBrackets.pop() ?? 0;
-					let end = index;
-					const part = value.substring(start + 1, end);
-					const level = openBrackets.length;
-					const decoration = parse(part, level);
-					if (decoration !== undefined) {
-						const rangeStart = offset + start + 1;
-						const rangeEnd = offset + end + 2;
-						colorMap.set(decoration, [...(colorMap.get(decoration) ?? []), new vscode.Range(editor.document.positionAt(rangeStart), editor.document.positionAt(rangeEnd))]);
-					}
-				}
+			colorPart(value, offset, colorMap, editor, true);
+		},
+		onObjectBegin: (offset: number, length: number, startLine: number, startCharacter: number, pathSupplier: () => JSONPath) => {
+			currentLevel++;
+		},
+		onObjectProperty: (property: string, offset: number, length: number, startLine: number, startCharacter: number, pathSupplier: () => JSONPath) => {
+			if (isInPlaceholders === currentLevel - 1) {
+				const rangeStart = offset + 1;
+				const rangeEnd = offset + property.length + 1;
+				const range = new vscode.Range(editor.document.positionAt(rangeStart), editor.document.positionAt(rangeEnd));
+				colorMap.set(argDecoration, [...(colorMap.get(argDecoration) ?? []), range]);
 			}
-		}
+			if (property.startsWith('@')) {
+				isInMetadata = currentLevel;
+			}
+			if (isInMetadata === currentLevel - 1 && property === "placeholders") {
+				isInPlaceholders = currentLevel;
+			}
+		},
+		onObjectEnd: (offset: number, length: number, startLine: number, startCharacter: number) => {
+			currentLevel--;
+			if (currentLevel < isInPlaceholders) {
+				isInPlaceholders = -1;
+			}
+			if (currentLevel < isInMetadata) {
+				isInMetadata = -1;
+			}
+		},
+	}, { disallowComments: true });
+	colorMap.forEach((value: vscode.Range[], key: vscode.TextEditorDecorationType) => {
+		editor.setDecorations(key, value ?? []);
 	});
-	editor.setDecorations(pluralDecoration, colorMap.get('plural') ?? []);
-	editor.setDecorations(selectDecoration, colorMap.get('select') ?? []);
-	editor.setDecorations(argDecoration, colorMap.get('arg') ?? []);
 }
 
-function parse(value: string, level: number): string | undefined {
-	if (selectRegex.exec(value) !== null) {
-		return 'select';
-	} else if (pluralRegex.exec(value) !== null) {
-		return 'plural';
-	} else {
-		return 'arg';
+function colorPart(value: string, offset: number, colorMap: Map<vscode.TextEditorDecorationType, vscode.Range[]>, editor: vscode.TextEditor, isOuter: boolean) {
+	const vals = XRegExp.matchRecursive(value, '\\{', '\\}', 'g');
```

Comment:
What happens if the braces are unbalanced (eg. because the user is typing)?

Reply:
also: does this format allow for the user to enter literal `{`'s with an escape character? (if so, will this handle them correctly?)

Reply:
Unbalanced brackets are handled gracefully, and now escape characters also work! Good catch.